fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/TypeCoercion.java

```diff
@@ -149,7 +150,8 @@ public Expression visitInPredicate(InPredicate inPredicate, ExpressionRewriteCon
         InPredicate newInPredicate = inPredicate.withChildren(rewrittenChildren);
 
         if (newInPredicate.getOptions().stream().map(Expression::getDataType)
-                .allMatch(dt -> dt.equals(newInPredicate.getCompareExpr().getDataType()))) {
+                .allMatch(dt -> dt.equals(NullType.INSTANCE)
+                        || dt.equals(newInPredicate.getCompareExpr().getDataType()))) {
             return newInPredicate;
         }
         Optional<DataType> optionalCommonType = TypeCoercionUtils.findWiderCommonType(newInPredicate.children()
```

Comment:
should also filter null type, when we do cast

Reply:
change castIfNotSameType() to castIfNotSameTypeAndNotNull().